### ISSUE TITLE
Feature Request: Return interface super type in scope child methods 

### ISSUE THREAD

**Issue by andreasnomikos** [uber/motif#52]
I encountered a problem while porting some of my Dagger modules to motif.
In Dagger when creating a component I can do something like 

```
@Component(
  modules = {FooDepsConsumerModule.class, BarDependenciesConsumerModule.class},
  dependencies = ParentComponent.class
)
class ChildScope implements BuilderComponent, FooDependencies, BarDependencies {
  
   static BuilderComponent createChildScope(ParentComponent parentComponent) {
     return Dagger_ChildScope.builder().build();
   }
}
```

this pattern is useful to limit the surface area that a consumer of ChildScope needs to rely on in terms of dependencies. e.g. otherwise the compiler would force resolution of all dependencies defined in FooDepsConsumerModule and BarDependenciesConsumerModule at the library that uses ChildScope.

I would like to be able to do something like the following in motif to emulate the same decomposition

```
@motif.Scope
interface ParentScope {
  
  @Child(ChildScope.class)  
  BuilderComponent childScope();
}
```

**Comment by Leland-Takamine** [comment]
Would wrapping in a class like you've done above with Dagger be a valid solution here?

**Comment by andreasnomikos** [comment]
I think it might work but seems a bit of an unnecessary overhead. Basically you are forced to write 2 motif scopes one for the upstream interface and 1 for the internal implementation.

**Comment by Leland-Takamine** [comment]
I'm trying to understand why the solution is any different when using Motif. Can you do this?

```java
@motif.Scope
interface Child extends BuilderComponent {
  static BuilderComponent createChildScope(ChildImpl.Dependencies dependencies) {
    return new ChildImpl(dependencies);
  }
}
```

**Comment by andreasnomikos** [comment]
Closed in favor of #56
